pkg/nmi/server/managed.go

```diff
@@ -0,0 +1,87 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	aadpodid "github.com/Azure/aad-pod-identity/pkg/apis/aadpodidentity"
+	auth "github.com/Azure/aad-pod-identity/pkg/auth"
+	k8s "github.com/Azure/aad-pod-identity/pkg/k8s"
+	"github.com/Azure/aad-pod-identity/pkg/nmi"
+	utils "github.com/Azure/aad-pod-identity/pkg/utils"
+	"github.com/Azure/go-autorest/autorest/adal"
+	"k8s.io/klog"
+)
+
+// ManagedClient implements the TokenClient interface
+type ManagedClient struct {
+	nmi.TokenClient
+	KubeClient   k8s.Client
+	IsNamespaced bool
+}
+
+// NewManagedTokenClient creates new managed token client
+func NewManagedTokenClient(client k8s.Client, isNamespaced bool) *ManagedClient {
+	return &ManagedClient{
+		KubeClient:   client,
+		IsNamespaced: isNamespaced,
+	}
+}
+
+// GetIdentities gets the azure identity that matches the podns/podname and client id
+func (mc *ManagedClient) GetIdentities(ctx context.Context, podns, podname, clientID string) (*aadpodid.AzureIdentity, error) {
+	// get pod object to retrieve labels
+	pod, err := mc.KubeClient.GetPod(podns, podname)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get pod %s/%s, err: %+v", podns, podname, err)
+	}
+	// get all the azure identities based on azure identity bindings
+	azureIdentities, err := mc.KubeClient.ListPodIdsWithBinding(podns, pod.Labels)
+
+	for _, id := range azureIdentities {
+		// if client id exists in the request, then send the first identity that matched the client id
+		if len(clientID) != 0 && id.Spec.ClientID == clientID {
+			return &id, nil
+		}
+		// if client doesn't exist in the request, then return the first identity in the same namespace as the pod
+		if len(clientID) == 0 && strings.EqualFold(id.Namespace, podns) {
+			klog.Infof("No clientID in request. %s/%s has been matched with azure identity %s/%s", podns, podname, id.Namespace, id.Name)
+			return &id, nil
+		}
+	}
+	return nil, fmt.Errorf("no matching azure identity found for pod")
+}
+
+// GetToken ...
+func (mc *ManagedClient) GetToken(ctx context.Context, rqClientID, rqResource string, podID aadpodid.AzureIdentity) (token *adal.Token, err error) {
+	rqHasClientID := len(rqClientID) != 0
+	clientID := podID.Spec.ClientID
+	if rqHasClientID && !strings.EqualFold(rqClientID, clientID) {
```

Comment:
Would this check for fail for Service Principal all the time since it would not have clientID ?

Reply:
or it might pass because both are empty, but not a required check for that code path ?